controllers/dashboard_controller.go

```diff
@@ -171,190 +166,210 @@ func (r *GrafanaDashboardReconciler) Reconcile(ctx context.Context, req ctrl.Req
 			}
 		}
 
 		// then import the dashboard into the matching grafana instances
 		err = r.onDashboardCreated(ctx, &grafana, cr, dashboardModel, hash)
 		if err != nil {
-			log.Error(err, "error reconciling dashboard", "dashboard", cr.Name, "grafana", grafana.Name)
 			applyErrors[fmt.Sprintf("%s/%s", grafana.Namespace, grafana.Name)] = err.Error()
-			success = false
 		}
 
-		condition := buildSynchronizedCondition("Dashboard", conditionDashboardSynchronized, cr.Generation, applyErrors, len(instances.Items))
-		meta.SetStatusCondition(&cr.Status.Conditions, condition)
-
 		if grafana.Spec.Preferences != nil && uid == grafana.Spec.Preferences.HomeDashboardUID {
 			err = r.UpdateHomeDashboard(ctx, grafana, uid, cr)
 			if err != nil {
-				return ctrl.Result{RequeueAfter: RequeueDelay}, err
+				applyHomeErrors[fmt.Sprintf("%s/%s", grafana.Namespace, grafana.Name)] = err.Error()
 			}
 		}
 	}
 
-	// if the dashboard was successfully synced in all instances, wait for its re-sync period
-	if success {
-		if cr.ResyncPeriodHasElapsed() {
-			cr.Status.LastResync = metav1.Time{Time: time.Now()}
-		}
-		cr.Status.Hash = hash
-		cr.Status.UID = uid
-		return ctrl.Result{RequeueAfter: cr.Spec.ResyncPeriod.Duration}, r.Client.Status().Update(ctx, cr)
+	// TODO Add new Condition displaying plugin reconciliation errors
+	if len(pluginErrors) > 0 {
+		err := fmt.Errorf("%v", pluginErrors)
+		log.Error(err, "failed to apply plugins to all instances")
+	}
+
+	// Specific to dashboards
+	// NOTE Add new Condition displaying home dash sync errors?
+	if len(applyHomeErrors) > 0 {
+		err := fmt.Errorf("%v", pluginErrors)
+		log.Error(err, "failed to apply home dashboards to all instances")
 	}
 
-	return ctrl.Result{RequeueAfter: RequeueDelay}, nil
+	condition := buildSynchronizedCondition("Dashboard", conditionDashboardSynchronized, cr.Generation, applyErrors, len(instances))
+	meta.SetStatusCondition(&cr.Status.Conditions, condition)
+
+	cr.Status.Hash = hash
+	cr.Status.UID = uid
+	return ctrl.Result{RequeueAfter: cr.Spec.ResyncPeriod.Duration}, nil
 }
 
-func (r *GrafanaDashboardReconciler) onDashboardDeleted(ctx context.Context, namespace string, name string) error {
+func (r *GrafanaDashboardReconciler) finalize(ctx context.Context, cr *v1beta1.GrafanaDashboard) error {
 	log := logf.FromContext(ctx)
-	list := v1beta1.GrafanaList{}
-	var opts []client.ListOption
-	err := r.Client.List(ctx, &list, opts...)
+	instances, err := GetScopedMatchingInstances(ctx, r.Client, cr)
 	if err != nil {
-		return err
+		return fmt.Errorf("fetching instances: %w", err)
 	}
 
-	for _, grafana := range list.Items {
-		if found, uid := grafana.Status.Dashboards.Find(namespace, name); found {
-			grafana := grafana
-			grafanaClient, err := client2.NewGeneratedGrafanaClient(ctx, r.Client, &grafana)
-			if err != nil {
-				return err
-			}
+	for _, grafana := range instances {
+		grafana := grafana
+		found, uid := grafana.Status.Dashboards.Find(cr.Namespace, cr.Name)
```

Comment:
Wouldn't it be simpler to unconditionally delete the dashboard from matching instances?

Reply:
With how much I'm already changing, I wanted to reduce the impact in logic as much as possible.
But since you're ok with it I will look into a few optimizations I have noted for the dashboard controller